### PR TITLE
Multi dim eval basis

### PR DESCRIPTION
**Pull request by BalzaniEdoardo**
This PR follows #274, additional edits to the basis for generalizing the behavior and removing the `__call__` method.

- Allow for N-dim inputs for both conv and eval basis
- Rename the `__call__` method of basis to `_evaluate`
- Make sure that the notebook exec error break the build (conf.py config)

EDIT: 

- I made the "set_kernel" method public because I am using it in the doc mostly
- The set_kernel will be removed from Eval basis in PR #276 (as discussed with @billbrod )

### CODE REVIEW

File: docs/background/basis/plot_01_1D_basis_function.md [hunk 1]
```diff
@@ -136,9 +136,55 @@ If you want to learn more about convolutions, as well as how and when to change
 check out the tutorial on [1D convolutions](convolution_background).
 :::
 
+### Multi-dimensional inputs
+For N-dimensional input with $N>1$, `compute_features` assumes the first axis represents samples. This is always valid for `pynapple` time series. For arrays, you can use [`numpy.transpose`](https://numpy.org/doc/stable/reference/generated/numpy.transpose.html) to re-arrange the axis if needed.
```
Comment (billbrod) [inline docs/background/basis/plot_01_1D_basis_function.md:140]:
```suggestion
For inputs with more than one dimension, `compute_features` assumes the first axis represents samples. This is always valid for `pynapple` time series. For arrays, you can use [`numpy.transpose`](https://numpy.org/doc/stable/reference/generated/numpy.transpose.html) to re-arrange the axis if needed.
```

File: docs/background/basis/plot_01_1D_basis_function.md [hunk 2]
```diff
@@ -136,9 +136,55 @@ If you want to learn more about convolutions, as well as how and when to change
...
+
+#### Eval Basis
+
+For Eval bases, `compute_features` evaluates the basis and then reshape the result into a 2D feature matrix.
```
Comment (billbrod) [inline docs/background/basis/plot_01_1D_basis_function.md:144]:
```suggestion
For Eval bases, `compute_features` evaluates the basis and outputs a 2D feature matrix.
```

File: docs/background/basis/plot_01_1D_basis_function.md [hunk 3]
```diff
@@ -136,9 +136,55 @@ If you want to learn more about convolutions, as well as how and when to change
...
+out = basis.compute_features(inp)
+out.shape
+```
+
+For each of the $3 \times 2 = 6$ inputs, `n_basis_funcs = 5` features are computed. These are concatenated on the second axis of the feature matrix, for a total of 
+$3 \times 2 \times 5  = 30$ outputs.
+
+#### Conv Basis
+
+For Conv bases, `compute_features` convolves each input with `n_basis_funcs` kernels, and reshaping the output into a 2D feature matrix.
```
Comment (billbrod) [inline docs/background/basis/plot_01_1D_basis_function.md:159]:
```suggestion
For Conv bases, `compute_features` convolves each input with `n_basis_funcs` kernels and outputs a 2D feature matrix.
```

File: docs/background/basis/plot_01_1D_basis_function.md [hunk 4]
```diff
@@ -136,9 +136,55 @@ If you want to learn more about convolutions, as well as how and when to change
...
+
+```{code-cell} ipython3
+basis = nmo.basis.RaisedCosineLinearConv(n_basis_funcs=5, window_size=6)
+# compute_features to perform the convolution and concatenate
+out = basis.compute_features(inp)
+print(f"`compute_features` output shape {out.shape}")
```
Comment (billbrod) [inline docs/background/basis/plot_01_1D_basis_function.md:165]:
```suggestion
out.shape
```
shoujld be consistent here and the Eval code block above. I don't have a preference for whether to print or not

File: docs/background/basis/plot_01_1D_basis_function.md [hunk 5]
```diff
@@ -136,9 +136,55 @@ If you want to learn more about convolutions, as well as how and when to change
...
+```
+
+This process is equivalent to performing the convolution separately with [`create_convolutional_predictor`](nemos.convolve.create_convolutional_predictor) and then reshaping the output.
+
+```{code-cell} ipython3
+# compute the kernels
```
Comment (billbrod) [inline docs/background/basis/plot_01_1D_basis_function.md:171]:
```suggestion
# cache the kernels
```

Comment (billbrod) [reply docs/background/basis/plot_01_1D_basis_function.md:171]: or something like "prepare" or "setup"

File: docs/background/basis/plot_01_1D_basis_function.md [hunk 6]
```diff
@@ -136,9 +136,55 @@ If you want to learn more about convolutions, as well as how and when to change
...
+For Conv bases, `compute_features` convolves each input with `n_basis_funcs` kernels, and reshaping the output into a 2D feature matrix.
+
+```{code-cell} ipython3
+basis = nmo.basis.RaisedCosineLinearConv(n_basis_funcs=5, window_size=6)
+# compute_features to perform the convolution and concatenate
+out = basis.compute_features(inp)
+print(f"`compute_features` output shape {out.shape}")
+```
+
+This process is equivalent to performing the convolution separately with [`create_convolutional_predictor`](nemos.convolve.create_convolutional_predictor) and then reshaping the output.
```
Comment (billbrod) [inline docs/background/basis/plot_01_1D_basis_function.md:168]: maybe this bit belongs in an admonition or collapsed block?

File: src/nemos/basis/_basis.py [hunk 7]
```diff
@@ -81,16 +81,16 @@ def min_max_rescale_samples(
...
-    # this passes if `samples_pts` contains a single value
-    if vmin != vmax:
-        scaling = vmax - vmin
-        sample_pts /= scaling
-    else:
-        scaling = 1.0
+
+    # do not normalize if samples contain a single value
+    scaling = np.asarray(vmax - vmin)
+    scaling[scaling == 0] = 1.0
```
Comment (billbrod) [inline src/nemos/basis/_basis.py:92]:
```suggestion
    scaling = np.asarray(vmax - vmin)
    # do not normalize if samples contain a single value (in which case vmax=vmin)
    scaling[scaling == 0] = 1.0
```
just to be more explicit

File: src/nemos/basis/_decaying_exponential.py [hunk 8]
```diff
@@ -156,23 +151,49 @@ def _evaluate(
...
-        )
-        # count the linear independent components (could be lower than n_basis_funcs for num precision).
-        n_independent_component = np.linalg.matrix_rank(exp_decay_eval)
-        # initialize output to nan
-        basis_funcs = np.full(
-            shape=(sample_pts.shape[0], n_independent_component), fill_value=np.nan
+
+        # process one input at the time (orthogonalization must be done one input at the time)
+
+        # flatten all inputs
```
Comment (billbrod) [inline src/nemos/basis/_decaying_exponential.py:157]:
```suggestion
        # flatten all non-time dimensions
```

**Review by billbrod — APPROVED** [review]
Just some small tweaks in wording, but otherwise this looks good to go! It's good how straightforward a change this was.

File: docs/background/basis/plot_01_1D_basis_function.md (hunk 5 above)
Comment (BalzaniEdoardo) [reply docs/background/basis/plot_01_1D_basis_function.md:171]: I like setup, it's more accessible :)

File: docs/background/basis/plot_01_1D_basis_function.md (hunk 6 above)
Comment (BalzaniEdoardo) [reply docs/background/basis/plot_01_1D_basis_function.md:168]: I like the idea of a collapsed block

**Pull request merged**